src/ast/passes/semantic_analyser.cpp

```diff
@@ -685,7 +685,27 @@ void SemanticAnalyser::visit(Call &call)
             << call.func << "() expects an integer or a pointer type as first "
             << "argument (" << t << " provided)";
       }
-      call.type = CreateString(bpftrace_.config_.get(ConfigKeyInt::max_strlen));
+
+      auto strlen = bpftrace_.config_.get(ConfigKeyInt::max_strlen);
+
+      if (call.vargs.size() == 2 && check_arg(call, Type::integer, 1, false)) {
+        auto &size_arg = *call.vargs.at(1);
+        if (size_arg.is_literal) {
+          auto &integer = static_cast<Integer &>(size_arg);
+          long value = integer.n;
+          if (value < 0) {
+            if (is_final_pass())
+              LOG(ERROR, call.loc, err_)
+                  << call.func << "cannot use negative length (" << value
+                  << ")";
+          } else {
+            strlen = value;
```

Comment:
Do we want the value to include NUL terminator or exclude?

Eg, if `str($1, 4)` is called, should we size a 4 byte buffer or 5?

Reply:
Should also probably check if `value` is bigger than max_strlen

Reply:
> Do we want the value to include NUL terminator or exclude?

Doesn't this happen automatically in codegen: https://github.com/bpftrace/bpftrace/blob/master/src/ast/passes/codegen_llvm.cpp#L782

Reply:
Yes, kinda. The generated conditional is mostly to handle dynamic len args. But if user provides a literal one I think it'd be better to error out. Principle of least surprise IMO